crates/sui-core/src/authority_server.rs

```diff
@@ -1010,6 +1063,136 @@ impl ValidatorService {
+            return Err(SuiError::UnsupportedFeatureError {
+                error: "Mysticeti fastpath".to_string(),
+            });
+        };
+        if request.epoch != epoch_store.epoch() {
+            return Err(SuiError::WrongEpoch {
+                expected_epoch: epoch_store.epoch(),
+                actual_epoch: request.epoch,
+            });
+        }
+        // Because we need to associate effects with a specific transaction position,
+        // we need to first make sure that this specific position is accepted by consensus,
+        // either with fastpath certified or post-commit finalized.
+        let first_status = consensus_tx_status_cache
+            .notify_read_transaction_status(request.transaction_position, None)
+            .await;
+        debug!(
+            tx_digest = ?request.transaction_digest,
+            "Observed consensus transaction status: {:?}",
+            first_status
+        );
+        let mut cur_status = match first_status {
+            NotifyReadConsensusTxStatusResult::Status(status) => match status {
+                ConsensusTxStatus::Rejected => {
+                    let response = WaitForEffectsResponse::Rejected {
+                        // TODO(fastpath): Add reject reason.
+                        reason: RejectReason::None,
+                    };
+                    return Ok(response);
+                }
+                ConsensusTxStatus::FastpathCertified | ConsensusTxStatus::Finalized => status,
+            },
+            NotifyReadConsensusTxStatusResult::Expired(round) => {
+                return Ok(WaitForEffectsResponse::Expired(round));
+            }
+        };
+        // Now that we know the transaction position is accepted by consensus,
+        // we can wait for the effects to be executed.
+        // In the meantime, however, if the initial status is fastpath certified,
+        // it is still possible that the transaction is rejected post commit.
+        // So we need to keep checking the status until it is finalized.
+        let effects = loop {
+            let transactions = [request.transaction_digest];
+            tokio::select! {
+                second_status = consensus_tx_status_cache.notify_read_transaction_status(request.transaction_position, Some(cur_status)) => {
+                    debug!(
+                        tx_digest = ?request.transaction_digest,
+                        "Observed consensus transaction status: {:?}",
+                        second_status
+                    );
+                    match second_status {
+                        NotifyReadConsensusTxStatusResult::Status(status) => {
+                            if status == ConsensusTxStatus::Rejected {
+                                return Ok(WaitForEffectsResponse::Rejected { reason: RejectReason::None });
+                            }
+                            assert!(matches!(status, ConsensusTxStatus::Finalized));
```

Comment:
We should be able to return effects if status is either `Finalized` or `FastpathCertified`. This is why the executed status needs the `finalized` bool

Reply:
We are though. This loop is tokio select waiting for the second status and the effects at the same time. If the effects come out first it will just return it.

Reply:
Yup my mistake, I see it now. Also after thinking about it we don't actually need the finalized bool. The client won't be able to do anything with this information. It has to wait for a quorum of effects regardless. Initially I thought if we got back effects with finalized status we could try and query for effects without consensus position but its probably not worth it.